File: RabinKarp.java
public class RabinKarp{
    String s;
    long[] prefix_hash;
    long[] powers;

    final int PRIME1 = 3137;
    final int PRIME2 = 1009;
    final int MOD1 = 1000000007;
    final int MOD2 = 998244353;

    /**
    * Class constructor specifiying the string s we want to work with.
    */
    RabinKarp(String s){
        this.s = s;
        this.prefix_hash = new long[this.s.length() + 1];
        this.powers = new long[this.s.length() + 1];
        this.computeHashAndPow();
    }

    private void computeHashAndPow(){
        this.powers[0] = 1;

        for(int i=1; i<=this.s.length(); i++){
            long c = (long) this.s.charAt(i-1) - 'a' + 1;
            this.prefix_hash[i] = ((prefix_hash[i-1]*PRIME1 + c) % MOD1);
            this.powers[i] = ((this.powers[i-1]*PRIME1) % MOD1);
        }
    }

    /**
    * Takes a left and right inclusive indices that resembles a substring and calculates the hash in O(1) time 
    * @param l The left of the substring
    * @param r The right of the substring
    * @return The hash of the substring [l,r].
    */
    public long getHashSubstring(int l, int r){
        return (prefix_hash[r+1] - prefix_hash[l] * powers[r - l + 1] % MOD1 + MOD1) % MOD1;
    }
}

File: RabinKarp.py
class RabinKarp:
    '''
    The RabinKarp object takes in a string s and is used to calculate the hash of any substring in O(1)
    :param s: The string we want to RabinKarp
    :type s: str

    ex: 
    rb = RabinKarp("hihi")
    hash1 = rb.getHashSubstring(0,1)
    hash2 = rb.getHashSubstring(2,3)
    // hash1 == hash2 
    '''
    PRIME1 = 3137
    PRIME2 = 1009
    MOD1 = 1000000007
    MOD2 = 998244353
    
    def __init__(self, s: str):
        self._s = s
        self._prefixHash = [0] * (len(s) + 1)
        self._powers = [0] * (len(s) + 1)
        self._computeHashAndPow()

    def _computeHashAndPow(self):
        self._powers[0] = 1

        for i in range(1, len(self._s) + 1):
            c = ord(self._s[i - 1]) - ord("a") + 1
            self._prefixHash[i] = (self._prefixHash[i - 1] * self.PRIME1 + c) % self.MOD1
            self._powers[i] = (self._powers[i - 1] * self.PRIME1) % self.MOD1

    def getHashSubstring(self, left: int, right: int):
        '''Calculates in O(1) the hash of the substring[left:right] INCLUSIVE

        :param left: The left index of the substring inclusive
        :type left: int
        :param right: The right index of the substring inclusive
        :type right: int

        :returns: An int of the hash of the substring
        :rtype: int
        '''  
        return (
            self._prefixHash[right + 1]
            - self._prefixHash[left] * self._powers[right - left + 1]
            + self.MOD1
        ) % self.MOD1
